### PR TITLE
HHH-11942 - ANTLR parser should fail when providing an extra parenthesis

### PR DESCRIPTION
**Pull request by vladmihalcea**
https://hibernate.atlassian.net/browse/HHH-11942

### CODE REVIEW

**Comment by vladmihalcea** [comment]
Fix issue by parsing the whole statements until the EOF instead of stopping at any unexpected right closing paranthesis.

**Comment by michael-simons** [comment]
Nice! Good improvement.

File: hibernate-core/src/test/java/org/hibernate/test/hql/ASTParserLoadingTest.java [hunk 1]
```diff
@@ -3786,19 +3786,23 @@ public void testEJBQLFunctions() throws Exception {
...
-		} );
+		try {
+			doInHibernate( this::sessionFactory, session -> {
+				session.createQuery(
+					"select a from Product a " +
+					"where " +
+					"coalesce(a.description, :description) = :description ) " +
+					"order by a.description ", Product.class)
+				.setParameter( "description", "desc" )
+				.getResultList();
```
Comment (dreab8) [inline hibernate-core/src/test/java/org/hibernate/test/hql/ASTParserLoadingTest.java]: I think a fail(); should be added

File: hibernate-core/src/test/java/org/hibernate/test/sql/function/JpaFunctionTest.java [hunk 2]
```diff
@@ -80,7 +81,7 @@ public void testWithoutCommaFail() {
 			} );
 		}
 		catch ( Exception e ) {
-			assertEquals( QuerySyntaxException.class, e.getCause().getClass() );
+			assertEquals( QueryException.class, e.getCause().getClass() );
```
Comment (dreab8) [inline hibernate-core/src/test/java/org/hibernate/test/sql/function/JpaFunctionTest.java]: in `o.h.hql.internal.ast. parse(boolean filter)` I would add
```
if ( hqlAst == null ) {
   parser.getParseErrorHandler().throwQueryException();
} 
```
after 
`final AST hqlAst = parser.getAST();`
this should avoid the change from `QuerySyntaxException` and `QueryException` and probably a better error cause message.

With this change the ASTParserLoadingTest# testOrderByExtraParenthesis () catch clause will become
	```
catch (IllegalArgumentException e) {
			final Throwable cause = e.getCause();
			assertThat( cause, instanceOf( QuerySyntaxException.class ) );
			assertTrue( cause.getMessage().contains( "expecting EOF, found ')'" ) );
		}
```

**Comment by dreab8** [comment]
here a PR https://github.com/hibernate/hibernate-orm/pull/1999 with the changes I proposed

**Comment by vladmihalcea** [comment]
You can close this one and let's use your new PR.

**Comment by dreab8** [comment]
superseded by  https://github.com/hibernate/hibernate-orm/pull/1999

**Pull request closed without merging**